File: src/com/rian/difficultycalculator/beatmap/BeatmapControlPointsManager.java
package com.rian.difficultycalculator.beatmap;

import com.rian.difficultycalculator.beatmap.timings.DifficultyControlPointManager;
import com.rian.difficultycalculator.beatmap.timings.EffectControlPointManager;
import com.rian.difficultycalculator.beatmap.timings.SampleControlPointManager;
import com.rian.difficultycalculator.beatmap.timings.TimingControlPointManager;

/**
 * A manager for beatmap control points.
 */
public class BeatmapControlPointsManager {
    /**
     * The manager for timing control points of this beatmap.
     */
    public final TimingControlPointManager timing;

    /**
     * The manager for difficulty control points of this beatmap.
     */
    public final DifficultyControlPointManager difficulty;

    /**
     * The manager for effect control points of this beatmap.
     */
    public final EffectControlPointManager effect;

    /**
     * The manager for sample control points of this beatmap.
     */
    public final SampleControlPointManager sample;

    public BeatmapControlPointsManager() {
        timing = new TimingControlPointManager();
        difficulty = new DifficultyControlPointManager();
        effect = new EffectControlPointManager();
        sample = new SampleControlPointManager();
    }

    /**
     * Copy constructor.
     *
     * @param source The source to copy from.
     */
    private BeatmapControlPointsManager(BeatmapControlPointsManager source) {
        timing = source.timing.deepClone();
        difficulty = source.difficulty.deepClone();
        effect = source.effect.deepClone();
        sample = source.sample.deepClone();
    }

    /**
     * Deep clones this manager.
     *
     * @return The deep cloned manager.
     */
    public BeatmapControlPointsManager deepClone() {
        return new BeatmapControlPointsManager(this);
    }
}

File: src/com/rian/difficultycalculator/beatmap/BeatmapControlPointsManager.kt
package com.rian.difficultycalculator.beatmap

import com.rian.difficultycalculator.beatmap.timings.DifficultyControlPointManager
import com.rian.difficultycalculator.beatmap.timings.TimingControlPointManager

/**
 * A manager for beatmap control points.
 */
class BeatmapControlPointsManager : Cloneable {
    /**
     * The manager for timing control points of this beatmap.
     */
    var timing = TimingControlPointManager()
        private set

    /**
     * The manager for difficulty control points of this beatmap.
     */
    var difficulty = DifficultyControlPointManager()
        private set

    override fun clone() =
        (super.clone() as BeatmapControlPointsManager).apply {
            timing = this@BeatmapControlPointsManager.timing.clone()
            difficulty = this@BeatmapControlPointsManager.difficulty.clone()
        }
}
